src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAPIUsageAnalyzer.cs

```diff
@@ -0,0 +1,200 @@
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DoNotUseTypesFromAssemblyDirectRule, DoNotUseTypesFromAssemblyIndirectRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                if (compilationStartContext.Compilation.GetTypeByMetadataName(CodeActionMetadataName) == null)
+                {
+                    // No reference to core Workspaces assembly.
+                    return;
+                }
+
+                INamedTypeSymbol diagnosticAnalyzer = compilationStartContext.Compilation.GetTypeByMetadataName(DiagnosticAnalyzerCorrectnessAnalyzer.DiagnosticAnalyzerTypeFullName);
+                if (diagnosticAnalyzer == null)
+                {
+                    // Does not contain any diagnostic analyzers.
+                    return;
+                }
+
+                var hasAccessToTypeFromWorkspaceAssemblies = false;
+                var namedTypesToAccessedTypesMap = new ConcurrentDictionary<INamedTypeSymbol, ImmutableHashSet<INamedTypeSymbol>>();
+                var diagnosticAnalyzerTypes = new ConcurrentBag<INamedTypeSymbol>();
+                compilationStartContext.RegisterSymbolAction(symbolContext =>
+                {
+                    var namedType = (INamedTypeSymbol)symbolContext.Symbol;
+                    if (namedType.DerivesFrom(diagnosticAnalyzer, baseTypesOnly: true))
+                    {
+                        diagnosticAnalyzerTypes.Add(namedType);
+                    }
+
+                    var usedTypes = GetUsedNamedTypes(namedType, symbolContext.Compilation, symbolContext.CancellationToken, ref hasAccessToTypeFromWorkspaceAssemblies);
+                    var added = namedTypesToAccessedTypesMap.TryAdd(namedType, usedTypes);
+                    Debug.Assert(added);
+                }, SymbolKind.NamedType);
+
+                compilationStartContext.RegisterCompilationEndAction(compilationEndContext =>
+                {
+                    if (diagnosticAnalyzerTypes.IsEmpty || !hasAccessToTypeFromWorkspaceAssemblies)
+                    {
+                        return;
+                    }
+
+                    var typesToProcess = new Queue<INamedTypeSymbol>();
+                    var processedTypes = new HashSet<INamedTypeSymbol>();
+                    var violatingTypeNamesBuilder = new SortedSet<string>();
+                    var violatingUsedTypeNamesBuilder = new SortedSet<string>();
+                    foreach (INamedTypeSymbol declaredType in namedTypesToAccessedTypesMap.Keys)
+                    {
+                        if (!diagnosticAnalyzerTypes.Contains(declaredType))
```

Comment:
is this possible?

Reply:
Yes, for helper types defined in assembly that are not deriving from `DiagnosticAnalyzer`, but may be accessed from diagnostic analyzers. These types might have helper methods that access types from Workspaces assemblies.